packages/wrapper/src/index.tsx

```diff
@@ -28,315 +27,354 @@ import {useRouter} from 'next/router';
+         */
+        if (config.debug) {
+            console.log('3. useHydration sync');
         }
 
-        return {store, props: {...initialProps, ...resultProps}};
-    };
-
-    const withRedux = (Component: NextComponentType | App | any) => {
-        console.warn(
-            '/!\\ You are using legacy implementation. Please update your code: use createWrapper() and wrapper.useWrappedStore().',
-        );
-
-        //TODO Check if pages/_app was wrapped so there's no need to wrap a page itself
-        const WrappedComponent = (props: any) => {
-            const {store, props: combinedProps} = useWrappedStore(props, WrappedComponent.displayName);
-
-            return (
-                <Provider store={store}>
-                    <Component {...combinedProps} />
-                </Provider>
-            );
-        };
-
-        WrappedComponent.displayName = `withRedux(${Component.displayName || Component.name || 'Component'})`;
+        hydrate();
 
-        if ('getInitialProps' in Component) {
-            WrappedComponent.getInitialProps = Component.getInitialProps;
+        if (config.debug) {
+            console.log('4. useHydration done');
         }
 
-        return WrappedComponent;
+        return {hydrating: false};
     };
 
     return {
         getServerSideProps,
         getStaticProps,
         getInitialAppProps,
         getInitialPageProps,
-        withRedux,
-        useWrappedStore,
+        useStore,
+        useHydration,
     };
 };
 
-// Legacy
-// eslint-disable-next-line import/no-anonymous-default-export
-export default <S extends Store>(makeStore: MakeStore<S>, config: Config<S> = {}) => {
-    console.warn('/!\\ You are using legacy implementation. Please update your code: use createWrapper() and wrapper.withRedux().');
-    return createWrapper(makeStore, config).withRedux;
-};
-
 export type Context = NextPageContext | AppContext | GetStaticPropsContext | GetServerSidePropsContext;
 
 export interface Config<S extends Store> {
-    serializeState?: (state: ReturnType<S['getState']>) => any;
-    deserializeState?: (state: any) => ReturnType<S['getState']>;
+    serializeAction?: (state: ReturnType<S['getState']>) => any;
+    deserializeAction?: (state: any) => ReturnType<S['getState']>;
     debug?: boolean;
+    actionFilter?: (action: Action) => boolean | any;
 }
 
-export interface WrapperProps {
-    initialProps: any; // stuff returned from getInitialProps or getServerSideProps
-    initialState: any; // stuff in the Store state after getInitialProps
+export interface PageProps {
+    initialStateGIAP: Action[]; // stuff in the Store state after App.getInitialProps
+    initialStateGIPP: Action[]; // stuff in the Store state after Page.getInitialProps
+    initialStateGSSP: Action[]; // stuff in the Store state after getServerSideProps
+    initialStateGSP: Action[]; // stuff in the Store state after getStaticProps
```

Comment:
I think these should all be optional?
```suggestion
    initialStateGIAP?: Action[]; // stuff in the Store state after App.getInitialProps
    initialStateGIPP?: Action[]; // stuff in the Store state after Page.getInitialProps
    initialStateGSSP?: Action[]; // stuff in the Store state after getServerSideProps
    initialStateGSP?: Action[]; // stuff in the Store state after getStaticProps
```